providers/implementations/asymciphers/sm2_enc.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <crypto/sm2.h>
+#include "prov/providercommonerr.h"
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "prov/provider_util.h"
+
+static OSSL_FUNC_asym_cipher_newctx_fn sm2_newctx;
+static OSSL_FUNC_asym_cipher_encrypt_init_fn sm2_init;
+static OSSL_FUNC_asym_cipher_encrypt_fn sm2_asym_encrypt;
+static OSSL_FUNC_asym_cipher_decrypt_init_fn sm2_init;
+static OSSL_FUNC_asym_cipher_decrypt_fn sm2_asym_decrypt;
+static OSSL_FUNC_asym_cipher_freectx_fn sm2_freectx;
+static OSSL_FUNC_asym_cipher_dupctx_fn sm2_dupctx;
+static OSSL_FUNC_asym_cipher_get_ctx_params_fn sm2_get_ctx_params;
+static OSSL_FUNC_asym_cipher_gettable_ctx_params_fn sm2_gettable_ctx_params;
+static OSSL_FUNC_asym_cipher_set_ctx_params_fn sm2_set_ctx_params;
+static OSSL_FUNC_asym_cipher_settable_ctx_params_fn sm2_settable_ctx_params;
+
+/*
+ * What's passed as an actual key is defined by the KEYMGMT interface.
+ * We happen to know that our KEYMGMT simply passes EC_KEY structures, so
+ * we use that here too.
+ */
+
+typedef struct {
+    OPENSSL_CTX *libctx;
+    EC_KEY *key;
+    PROV_DIGEST md;
+} PROV_SM2_CTX;
+
+static void *sm2_newctx(void *provctx)
+{
+    PROV_SM2_CTX *psm2ctx =  OPENSSL_zalloc(sizeof(PROV_SM2_CTX));
+
+    if (psm2ctx == NULL)
+        return NULL;
+    psm2ctx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+
+    return psm2ctx;
+}
+
+static int sm2_init(void *vpsm2ctx, void *vkey)
+{
+    PROV_SM2_CTX *psm2ctx = (PROV_SM2_CTX *)vpsm2ctx;
+
+    if (psm2ctx == NULL || vkey == NULL || !EC_KEY_up_ref(vkey))
+        return 0;
+    EC_KEY_free(psm2ctx->key);
+    psm2ctx->key = vkey;
+
+    return 1;
+}
+
+static const EVP_MD *sm2_get_md(PROV_SM2_CTX *psm2ctx)
+{
+    const EVP_MD *md = ossl_prov_digest_md(&psm2ctx->md);
+
+    if (md == NULL)
+        md = ossl_prov_digest_fetch(&psm2ctx->md, psm2ctx->libctx, "SM3", NULL);
```

Comment:
propq needs to be passed here..
Which means it should be stored in the struct and be settable.

Reply:
Does it? You can already provide properties and a digest name and they are settable. If you do then this code is never run because we already fetched the digest you specified using the properties you specified. This code kicks in where you haven't supplied an explicit digest name. So the only way we could get here and props be anything other than NULL is if you have not supplied an explicit digest name, but you did supply properties. I'm not sure that's a scenario we need to support??

Reply:
ok

Reply:
(this is to my knowledge hypothetical) would you ever want to just set the fips property but not a digest name?

Reply:
Not sure I understand what fips has to do with this issue? Which fips property do you mean?

Reply:
It's possible I'm just confused, in which case sorry for the disruption.
But I thought that one of the uses of the propq was to be able to set `fips=yes` to require the fips property to be set for the fetched algorithm (which in practice will usually mean from our fips provider, but in theory could be a different one).
It did not stretch credulity to think that someone might want to restrict the fetched digest to being a fips-approved one without specifying exactly which one it has to be.  From just the context of this thread it seemed like that might trigger the case in question ("have not supplied an explicit digest name, but you did supply properties").

Reply:
Right. The propq here should not be confused with the default propq that someone might use. So in order for this scenario to occur someone has explicitly sent a propq parameter to the sm2 asym cipher implementation, but not specified the digest. The assumption in this PR is that you must set both if you want the propq to be applied. So if you want to specify a specific propq then you must also specific a specific digest.

Reply:
Thanks for the extra explanation, and sorry for my confusion.  I think I get it now.